modules/imgcodecs/test/test_read_write.cpp

```diff
@@ -303,4 +303,142 @@ TEST(Imgcodecs_Image, write_umat)
+    };
+
+    ImageCollection collection(filename, IMREAD_ANYCOLOR);
+    auto firstIter = collection.begin();
+    auto secondIter = collection.begin();
+
+    // Decode all odd pages then decode even pages -> 1, 0, 3, 2 ...
+    firstIter++;
+    for(size_t i = 1; i < collection.size(); i += 2, ++firstIter, ++firstIter, ++secondIter, ++secondIter) {
+        Mat mat = *firstIter;
+        double diff = cv::norm(mat, imread(page_files[i]), NORM_INF);
+        EXPECT_EQ(0., diff);
+        Mat evenMat = *secondIter;
+        diff = cv::norm(evenMat, imread(page_files[i-1]), NORM_INF);
+        EXPECT_EQ(0., diff);
+    }
+}
+
+TEST(Imgcodecs_Image, multipage_collection_operator_plusplus)
+{
+    const string root = cvtest::TS::ptr()->get_data_path();
+    const string filename = root + "readwrite/multipage.tif";
+
+    // operator++ test
+    ImageCollection collection(filename, IMREAD_ANYCOLOR);
+    auto firstIter = collection.begin();
+    auto secondIter = firstIter++;
+
+    // firstIter points to second page, secondIter points to first page
+    double diff = cv::norm(*firstIter, *secondIter, NORM_INF);
+    EXPECT_NE(diff, 0.);
+}
+
+TEST(Imgcodecs_Image, multipage_collection_backward_decoding)
+{
+    const string root = cvtest::TS::ptr()->get_data_path();
+    const string filename = root + "readwrite/multipage.tif";
+    const string page_files[] = {
+            root + "readwrite/multipage_p1.tif",
+            root + "readwrite/multipage_p2.tif",
+            root + "readwrite/multipage_p3.tif",
+            root + "readwrite/multipage_p4.tif",
+            root + "readwrite/multipage_p5.tif",
+            root + "readwrite/multipage_p6.tif"
+    };
+
+    ImageCollection collection(filename, IMREAD_ANYCOLOR);
+    // backward decoding -> 5,4,3,2,1,0
+    for(int i = (int)collection.size() - 1; i >= 0; --i)
+    {
+        double diff = cv::norm(collection[i], imread(page_files[i]), NORM_INF);
+        EXPECT_EQ(diff, 0.);
+    }
+
+    for(int i = 0; i < (int)collection.size(); ++i)
+    {
+        collection.release(i);
+    }
+
+    double diff = cv::norm(collection[2], imread(page_files[2]), NORM_INF);
+    EXPECT_EQ(diff, 0.);
+}
+
+TEST(ImgCodecs, multipage_decoding_range_based_for_loop_test)
+{
+    const string root = cvtest::TS::ptr()->get_data_path();
+    const string filename = root + "readwrite/multipage.tif";
+    const string page_files[] = {
+            root + "readwrite/multipage_p1.tif",
+            root + "readwrite/multipage_p2.tif",
+            root + "readwrite/multipage_p3.tif",
+            root + "readwrite/multipage_p4.tif",
+            root + "readwrite/multipage_p5.tif",
+            root + "readwrite/multipage_p6.tif"
+    };
+
+    ImageCollection collection(filename, IMREAD_ANYCOLOR);
+
+    int index = 0;
+    for(auto &i: collection)
+    {
+        double diff = cv::norm(i, imread(page_files[index]), NORM_INF);
+        EXPECT_EQ(0., diff);
+        ++index;
+    }
```

Comment:
It makes sense to check index after the loop to ensure that all pages has been read.